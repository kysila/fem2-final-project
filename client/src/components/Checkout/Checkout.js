import React, { useState } from 'react';
import {
  Grid, Typography, Checkbox, FormControlLabel, Button, Divider, Box, useMediaQuery, withTheme,
} from '@material-ui/core';
import LocationOnIcon from '@material-ui/icons/LocationOn';
import ArrowBackIosIcon from '@material-ui/icons/ArrowBackIos';
import { connect } from 'react-redux';
import { Logo } from '../../commons/Logo/Logo';
import { Steps } from '../../commons/Steps/Steps';
import { InputField } from '../../commons/InputField/InputField';

import { useStyles } from './styles';


import MuseoSans from '../../fonts/MuseoSans-500.woff';
import { dispatchModalOpen } from '../../store/modal/actions';

const museo = {
  fontFamily: 'Museo Sans 500',
  fontStyle: 'normal',
  src: `
    local('Museo Sans 500'),
    url(${MuseoSans}) format('woff')
  `,
};

function CheckoutOrderProduct(props) {
  return (
    <Grid
      container
      direction="row"
      alignItems="center"
      justify="space-between"
      style={{ marginBottom: '20px' }}
    >
      <Grid
        item
        lg={3}
        md={3}
        xs={3}
      >
        <img width={64} height={64} style={{ marginRight: '10px' }} src={props.src} alt={props.title} />
      </Grid>
      <Grid
        item
        lg={9}
        md={9}
        xs={9}
      >
        <Grid
          container
          direction="column"
          alignItems="flex-start"
          justify="space-between"
        >
          <Typography style={{ fontSize: '12px', lineHeight: '20px', ...museo }}>
            {props.title}
          </Typography>
          <Grid
            container
            direction="row"
            alignItems="center"
            justify="space-between"
          >
            <Typography style={{
              color: '#888888', fontSize: '12px', lineHeight: '20px', ...museo,
            }}
            >
              { `Quantity: ${props.quantity}` }
            </Typography>
            <Typography style={{ fontSize: '12px', lineHeight: '20px', ...museo }}>
              $
              {props.price}
            </Typography>
          </Grid>
        </Grid>
      </Grid>
    </Grid>
  );
}

let Checkout = withTheme((props) => {
  const classes = useStyles(props);
  const tablet = useMediaQuery(props.theme.breakpoints.down(1024));

  const [state, setState] = useState({
    email: '',
    phone: '',
    firstName: '',
    lastName: '',
    company: '',
    appartment: '',
    address: '',
    promote: true,
    discount: '',
  });

  const mockProducts = [
    {
      key: 'id-0',
      src: '/img/products/e-bikes/872426/001.jpg',
      title: 'Addmotor Hithot H1 Sport Mountain E-bike (green)',
      quantity: 1,
      price: '1699.99',
    },
    {
      key: 'id-1',
      src: '/img/products/e-bikes/872426/001.jpg',
      title: 'Civi Bikes Cheetah Vintage-Style Electric Fat Bike (Night Black)',
      quantity: 1,
      price: '2299.99',
    },
  ];

  const steps = [
    {
      key: 'id-0',
      isActive: true,
      text: 'Cart',
    },
    {
      key: 'id-1',
      isActive: true,
      text: 'Information',
    },
    {
      key: 'id-2',
      isActive: false,
      text: 'Shipping',
    },
    {
      key: 'id-3',
      isActive: false,
      text: 'Payment',
    },
  ];

  const onChange = (stateName) => (event) => {
    setState({ ...state, [stateName]: event.target.checked || event.target.value });
  };

  return (
    <Grid
      container
      direction="row"
      alignItems="center"
      justify="center"
      className={classes.wrapper}
    >
      <Grid
        className={classes.container}
        container
        direction="column"
        alignItems="flex-start"
        justify="flex-start"
      >
        <Logo />
        <Steps className={classes.steps} items={steps} />
        <Grid
          className={classes.body}
          container
          direction={tablet ? 'column-reverse' : 'row'}
          alignItems={tablet ? 'center' : 'flex-start'}
          justify="center"
        >
          {
            tablet ? (
              <Grid />
            ) : null
          }
          <Grid
            item
            lg={tablet ? 12 : 7}
            md={tablet ? 12 : 7}
            xs={tablet ? 12 : 7}
          >
            <Grid
              className={classes.contact}
              container
              direction="row"
              alignItems="center"
              justify="center"
            >
              <Grid
                container
                direction="column"
                alignItems="center"
                justify="center"
              >
                <Grid
                  container
                  direction="row"
                  alignItems="center"
                  justify="space-between"
                >
                  <Typography style={{ lineHeight: '30px', fontSize: '20px', ...museo }}>
                    Contact information
                  </Typography>
                  <Typography style={{ fontSize: '14px', ...museo }}>
                    Already have an account?
                    <Typography component="span" style={{ color: '#8F8DE2', cursor: 'pointer' }} onClick={props.openLogin}> Log In</Typography>
                  </Typography>
                </Grid>
                <Grid
                  container
                  direction="row"
                  alignItems="center"
                  justify="center"
                  spacing={2}
                >
                  <Grid
                    item
                    lg={6}
                    md={6}
                    xs={6}
                  >
                    <InputField
                      id="email"
                      type="email"
                      label="Email"
                      labelWidth={35}
                      onChange={onChange('email')}
                      value={state.email}
                    />
                  </Grid>
                  <Grid
                    item
                    lg={6}
                    md={6}
                    xs={6}
                  >
                    <InputField
                      id="phone"
                      type="tel"
                      label="Phone"
                      labelWidth={40}
                      onChange={onChange('phone')}
                      value={state.phone}
                    />
                  </Grid>
                  <Grid
                    item
                    lg={12}
                    md={12}
                    xs={12}
                  >
                    <FormControlLabel
                      control={<Checkbox color="default" checked={state.promote} onChange={onChange('promote')} value="" />}
                      label="Keep me up to date on news and exclusive offers"
                    />
                  </Grid>
                </Grid>
                <Grid
                  className={classes.shipping}
                  container
                  direction="row"
                  alignItems="center"
                  justify="space-between"
                >
                  <Typography style={{ lineHeight: '30px', fontSize: '20px', ...museo }}>
                    Shipping address
                  </Typography>
                </Grid>
                <Grid
                  container
                  direction="row"
                  alignItems="center"
                  justify="center"
                  spacing={2}
                >
                  <Grid
                    item
                    lg={6}
                    md={6}
                    xs={6}
                  >
                    <InputField
                      id="firstName"
                      type="text"
                      label="First name"
                      labelWidth={65}
                      onChange={onChange('firstName')}
                      value={state.firstName}
                    />
                  </Grid>
                  <Grid
                    item
                    lg={6}
                    md={6}
                    xs={6}
                  >
                    <InputField
                      id="lastName"
                      type="text"
                      label="Last name"
                      labelWidth={65}
                      onChange={onChange('lastName')}
                      value={state.lastName}
                    />
                  </Grid>
                </Grid>
                <Grid
                  container
                  direction="row"
                  alignItems="center"
                  justify="center"
                  spacing={2}
                >
                  <Grid
                    item
                    lg={6}
                    md={6}
                    xs={6}
                  >
                    <InputField
                      id="company"
                      type="text"
                      label="Company (optional)"
                      labelWidth={130}
                      onChange={onChange('company')}
                      value={state.company}
                    />
                  </Grid>
                  <Grid
                    item
                    lg={6}
                    md={6}
                    xs={6}
                  >
                    <InputField
                      id="appartment"
                      type="text"
                      label="Appartment, suite, etc (optional)"
                      labelWidth={210}
                      onChange={onChange('appartment')}
                      value={state.appartment}
                    />
                  </Grid>
                </Grid>
                <Grid
                  container
                  direction="row"
                  alignItems="center"
                  justify="center"
                  spacing={2}
                >
                  <Grid
                    item
                    lg={12}
                    md={12}
                    xs={12}
                  >
                    <InputField
                      id="address"
                      type="text"
                      label="Address (start writing or select on the map)"
                      labelWidth={280}
                      onChange={onChange('company')}
                      value={state.address}
                      InputProps={{
                        endAdornment: (
                          <LocationOnIcon
                            style={{ color: '#888888' }}
                            onClick={() => props.openMap({ onChange: onChange('address'), className: classes.mapModal })}
                          />),
                      }}
                    />
                  </Grid>
                </Grid>
                <Grid
                  className={classes.navigation}
                  container
                  direction="row"
                  alignItems="center"
                  justify="space-between"
                >
                  <Grid
                    item
                    lg={6}
                    md={6}
                    xs={6}
                  >
                    <Typography component="span" className={classes.editCart} style={museo}>
                      <ArrowBackIosIcon style={{ fontSize: '14px', lineHeight: '30px' }} />
                      Edit Cart
                    </Typography>
                  </Grid>
                  <Grid
                    item
                    lg={6}
                    md={6}
                    xs={6}
                  >
                    <Grid
                      container
                      direction="row"
                      alignItems="center"
                      justify="flex-end"
                    >
                      <Button>Continue to Shipping</Button>
                    </Grid>
                  </Grid>
                </Grid>
              </Grid>
            </Grid>
          </Grid>
          <Grid
            item
            lg={1}
            md={1}
            xs={1}
          />
          {
            !tablet ? (
              <Grid
                item
                lg={4}
                md={4}
                xs={4}
              >
                <Grid
                  container
                  direction="column"
                  alignItems="flex-start"
                  justify="flex-start"
                >
                  <Typography
                    style={{
                      fontSize: '16px',
                      lineHeight: '30px',
                      color: '#888888',
                      ...museo,
                    }}
                  >
                  Order Summary
                  </Typography>
                </Grid>
                {mockProducts.map((item) => <CheckoutOrderProduct {...item} />)}
                <Divider className={classes.divider} />
                <InputField
                  id="discount"
                  type="text"
                  label="Discount code"
                  labelWidth={110}
                  value={state.discount}
                  onChange={onChange('discount')}
                  FormControlProps={{
                    className: classes.dividerFormControl,
                  }}
                  LabelProps={{
                    classes: {
                      root: classes.discountLabel,
                    },
                  }}
                  InputProps={{
                    classes: {
                      input: classes.discountInput,
                      adornedEnd: classes.discountAdornedEnd,
                    },
                    endAdornment: (
                      <Button
                        style={museo}
                        className={classes.discountButton}
                        variant="text"
                      >
                      Apply
                      </Button>
                    ),
                  }}
                />
                <Divider className={classes.divider} />
                <Grid
                  container
                  direction="column"
                  alignItems="center"
                  justify="center"
                >
                  <Grid
                    container
                    direction="row"
                    alignItems="center"
                    justify="space-between"
                  >
                    <Typography className={classes.afterDiscount} style={museo}>
                    Subtotal
                    </Typography>
                    <Typography className={classes.afterDiscount} style={museo}>
                    $3,999.98
                    </Typography>
                  </Grid>
                  <Grid
                    container
                    direction="row"
                    alignItems="center"
                    justify="space-between"
                  >
                    <Typography className={classes.afterDiscount} style={museo}>
                    Shipping
                    </Typography>
                    <Typography className={classes.afterDiscount} style={museo}>
                    Calculated at next step
                    </Typography>
                  </Grid>
                </Grid>
                <Divider className={classes.divider} />
                <Grid
                  container
                  direction="row"
                  alignItems="center"
                  justify="space-between"
                >
                  <Typography className={classes.total} style={museo}>
                  Total
                  </Typography>
                  <Typography className={classes.total} style={museo}>
                  $
                    {
                    mockProducts
                      .reduce((total, { price }) => (Number(total) + Number(price)).toFixed(2), 0)
                  }
                  </Typography>
                </Grid>
              </Grid>
            ) : null
          }
        </Grid>
      </Grid>
      <Box
        className={classes.footer}
      >
        <Typography style={{ color: '#888888', fontSize: '12px', ...museo }}>
          Copyright © 2019 Electra.
        </Typography>
      </Box>
    </Grid>
  );
});

Checkout = connect(null, {
  openMap: (inject) => dispatchModalOpen('map', inject),
})(Checkout);

export {
  Checkout,
  CheckoutOrderProduct,
};
