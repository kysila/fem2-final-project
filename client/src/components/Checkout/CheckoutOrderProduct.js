import React from 'react';
import {
  Grid, Typography, useMediaQuery, withTheme,
} from '@material-ui/core';
import MuseoSans from '../../fonts/MuseoSans-500.woff';

const museo = {
  fontFamily: 'Museo Sans 500',
  fontStyle: 'normal',
  src: `
    local('Museo Sans 500'),
    url(${MuseoSans}) format('woff')
  `,
};

export const CheckoutOrderProduct = withTheme((props) => {
  const tablet = useMediaQuery(props.theme.breakpoints.down(1024));
  const mobile = useMediaQuery(props.theme.breakpoints.down(768));

  return (
    <Grid
      container
      direction="row"
      alignItems="center"
      justify="space-between"
      style={{ marginBottom: '20px' }}
    >
      <Grid
        item
        lg={tablet && !mobile ? 1 : 3}
        md={tablet && !mobile ? 1 : 3}
        xs={tablet && !mobile ? 1 : 3}
      >
        <img width={64} height={64} style={{ marginRight: '10px' }} src={props.src} alt={props.title} />
      </Grid>
      <Grid
        item
        lg={tablet && !mobile ? 11 : 9}
        md={tablet && !mobile ? 11 : 9}
        xs={tablet && !mobile ? 11 : 9}
      >
        <Grid
          container
          direction={tablet && !mobile ? 'row' : 'column'}
          alignItems={tablet && !mobile ? 'flex-start' : null}
          justify="space-between"
        >
          <Grid
            item
            lg={tablet && !mobile ? 7 : 12}
            md={tablet && !mobile ? 7 : 12}
            xs={tablet && !mobile ? 7 : 12}
          >
            <Grid
              container
              alignItems="flex-start"
              justify="flex-start"
            >
              <Typography style={{ fontSize: '12px', lineHeight: '20px', ...museo }}>
                {props.title}
              </Typography>
            </Grid>
          </Grid>
          <Grid
            item
            lg={tablet && !mobile ? 5 : 12}
            md={tablet && !mobile ? 5 : 12}
            xs={tablet && !mobile ? 5 : 12}
          >
            <Grid
              container
              alignItems="center"
              justify="space-between"
            >
              <Typography style={{
                color: '#888888', fontSize: '12px', lineHeight: '20px', ...museo,
              }}
              >
                { `Quantity: ${props.quantity}` }
              </Typography>
              <Typography style={{ fontSize: '12px', lineHeight: '20px', ...museo }}>
                $
                {props.price}
              </Typography>
            </Grid>
          </Grid>
        </Grid>
      </Grid>
    </Grid>
  );
});
