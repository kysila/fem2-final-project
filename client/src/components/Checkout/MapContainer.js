import {
  Box, Button, Grid, Typography,
} from '@material-ui/core';
import CloseIcon from '@material-ui/icons/Close';
import React from 'react';
import { useStyles } from './styles';
import Tungsten from '../../fonts/Tungsten-Book.woff';
import MuseoSans from '../../fonts/MuseoSans-500.woff';

const tungsten = {
  fontFamily: 'Tungsten Book',
  fontStyle: 'normal',
  src: `
    local('Tungsten Book'),
    url(${Tungsten}) format('woff')
  `,
};

const museo = {
  fontFamily: 'Museo Sans 500',
  fontStyle: 'normal',
  src: `
    local('Museo Sans 500'),
    url(${MuseoSans}) format('woff')
  `,
};

export function MapContainer(props) {
  const classes = useStyles();
  return (
    <Box
      className={classes.mapContainer}
    >
      <Grid
        container
        direction="column"
        alignItems="center"
        justify="flex-start"
        className={classes.mapBody}
      >
        <Typography style={{ fontSize: '36px', ...tungsten }}>
          Location Selector
        </Typography>
        <Typography style={{
          marginTop: '8px', marginBottom: '20px', fontSize: '16px', color: '#888888', textAlign: 'center', ...museo,
        }}
        >
          Drag the map to place locator on the needed place.
          <br />
          When ready - press “Done”. You can then edit address inside Address field.
        </Typography>
        {props.children}
        <Button
          style={{
            padding: '13px 42px', fontSize: '14px', marginTop: '20px', fontWeight: 'bold', ...museo,
          }}
          onClick={props.onClick}
        >
          Done
        </Button>
      </Grid>
      <Box className={classes.closeWrapper} onClick={props.onClick}>
        <CloseIcon className={classes.close} />
      </Box>
    </Box>
  );
}
