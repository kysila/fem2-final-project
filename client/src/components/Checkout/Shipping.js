import React, { useEffect, useState } from 'react';
import {
  Button, Grid, Typography, Divider, Radio, FormControlLabel, useMediaQuery, withTheme,
} from '@material-ui/core';
import ArrowBackIosIcon from '@material-ui/icons/ArrowBackIos';
import { connect } from 'react-redux';
import { withRouter, Link } from 'react-router-dom';
import { useStyles } from './styles';
import MuseoSans from '../../fonts/MuseoSans-500.woff';
import { dispatchModalClose, dispatchModalOpen } from '../../store/modal/actions';
import { MapContainer } from './MapContainer';
import { ArrowTooltip } from '../../commons/Tooltip/Tooltip';

const museo = {
  fontFamily: 'Museo Sans 500',
  fontStyle: 'normal',
  src: `
    local('Museo Sans 500'),
    url(${MuseoSans}) format('woff')
  `,
};

export const Shipping = withRouter(connect(({ auth: { user } }) => ({ user }), {
  openMap: (inject) => dispatchModalOpen('map', inject),
  closeModal: () => dispatchModalClose(),
})(withTheme((props) => {
  const classes = useStyles();
  const mobile = useMediaQuery(props.theme.breakpoints.down(768));

  const [shippingMethod, setShippingMethod] = useState('free');
  const [info, setInfo] = useState(JSON.parse(localStorage.getItem('/checkout/shipping'))
    || JSON.parse(localStorage.getItem('/checkout/info'))
    || {
      email: { value: '' },
      address: { value: '', error: null },
    });

  const validate = () => {
    if (!info.address.value) {
      setInfo({ ...info, address: { ...info.address, error: 'Please fill ship address.' } });
      return false;
    }

    return true;
  };

  useEffect(() => {
    if (props.user) {
      setInfo({ ...info, email: { value: props.user.email } });
    }
    // eslint-disable-next-line
  }, [info.email.value, props.user]);

  const onSubmit = () => {
    if (validate()) {
      localStorage.setItem('/checkout/shipping', JSON.stringify(info));
      props.history.push('/checkout/payment');
    }
  };

  return (
    <Grid
      className={classes.contact}
      container
      direction="row"
      alignItems="center"
      justify="center"
      style={{ marginTop: '18px' }}
    >
      <Grid
        container
        direction="column"
        alignItems="center"
        justify="center"
      >
        <Grid
          container
          direction="row"
          alignItems="center"
          justify="center"
        >
          <Grid
            item
            lg={2}
            md={2}
            xs={2}
          >
            <Typography style={{ fontSize: '14px', color: '#888888', ...museo }}>
              Contact
            </Typography>
          </Grid>
          <Grid
            item
            lg={8}
            md={8}
            xs={8}
          >
            <Typography style={{ fontSize: '14px' }}>
              {info.email.value}
            </Typography>
          </Grid>
          <Grid
            item
            lg={2}
            md={2}
            xs={2}
          >
            <Link to="/checkout/info" style={{ fontSize: '14px', color: '#6A86E8', ...museo }}>
              Change
            </Link>
          </Grid>
        </Grid>
        <Grid
          container
          direction="row"
          alignItems="center"
          justify="center"
          style={{ marginTop: '10px' }}
        >
          <Grid
            item
            lg={2}
            md={2}
            xs={2}
          >
            <Typography style={{
              whiteSpace: 'nowrap', fontSize: '14px', color: '#888888', ...museo,
            }}
            >
              Ship to
            </Typography>
          </Grid>
          <Grid
            item
            lg={8}
            md={8}
            xs={8}
            style={{
              minWidth: '0',
            }}
          >
            <Typography style={{
              fontSize: '14px',
              textOverflow: 'ellipsis',
              whiteSpace: 'nowrap',
              overflow: 'hidden',
              ...museo,
            }}
            >
              {info.address.value}
            </Typography>
          </Grid>
          <ArrowTooltip title={info.address.error}>
            <Grid
              item
              lg={2}
              md={2}
              xs={2}
            >
              <Typography
                onClick={() => props.openMap({
                  onChange: ({ target: { value } }) => setInfo({ ...info, address: { value } }),
                  className: classes.map,
                  mapElementClass: classes.mapElement,
                  address: info.address.value.trim(),
                  marker: null,
                  Container: (pps) => <MapContainer {...pps} onClick={props.closeModal} />,
                })}
                style={{
                  fontSize: '14px', cursor: 'pointer', color: '#6A86E8', ...museo,
                }}
              >
                Change
              </Typography>
            </Grid>
          </ArrowTooltip>
        </Grid>
        <Divider style={{ width: '100%', margin: '20px 0' }} />
        <Grid
          className={classes.shipping}
          container
          direction={mobile ? 'column' : 'row'}
          alignItems={mobile ? null : 'center'}
          justify="space-between"
        >
          <Typography style={{ lineHeight: '30px', fontSize: '20px', ...museo }}>
            Shipping Method
          </Typography>
        </Grid>
        <Grid
          container
          direction="column"
          alignItems={mobile ? null : 'flex-start'}
          justify="space-between"
        >
          <FormControlLabel
            value="free"
            control={(
              <Radio
                checked={shippingMethod === 'free'}
                onChange={() => setShippingMethod('free')}
                value="free"
                color="default"
                name="free"
              />
            )}
            label={(
              <Grid container direction="column">
                <Typography style={{ fontSize: '16px', ...museo }}>
                  Free Shipping
                </Typography>
                <Typography style={{ fontSize: '14px', color: '#888888', ...museo }}>
                  Slow shipping to your destination address.
                </Typography>
              </Grid>
            )}
            labelPlacement="end"
          />
          <FormControlLabel
            value="high-cost"
            control={(
              <Radio
                checked={shippingMethod === 'high-cost'}
                onChange={() => setShippingMethod('high-cost')}
                value="high-cost"
                color="default"
                name="high-cost"
              />
            )}
            label={(
              <Grid container direction="column">
                <Typography style={{ fontSize: '16px', ...museo }}>
                  Super high-cost shipping
                </Typography>
                <Typography style={{ fontSize: '14px', color: '#888888', ...museo }}>
                  Really fast shipping to your destination address.
                </Typography>
              </Grid>
            )}
            labelPlacement="end"
            style={{ marginTop: '20px' }}
          />

        </Grid>
        <Grid
          className={classes.navigation}
          container
          direction={mobile ? 'column' : 'row'}
          alignItems="center"
          justify="space-between"
        >
          <Grid
            item
            lg={mobile ? 12 : 6}
            md={mobile ? 12 : 6}
            xs={mobile ? 12 : 6}
          >
            <Link to="/checkout/info">
              <Typography component="span" className={classes.editCart} style={museo}>
                <ArrowBackIosIcon style={{ fontSize: '14px', lineHeight: '30px' }} />
                Edit Information
              </Typography>
            </Link>
          </Grid>
          <Grid
            item
            lg={mobile ? 12 : 6}
            md={mobile ? 12 : 6}
            xs={mobile ? 12 : 6}
          >
            <Grid
              container
              direction="row"
              alignItems={mobile ? null : 'center'}
              justify={mobile ? 'center' : 'flex-end'}
            >
              <Button
                onClick={onSubmit}
                style={{
                  padding: '13px 42px', fontSize: '14px', marginTop: '20px', fontWeight: 'bold', ...museo,
                }}
              >
                Continue to Payment
              </Button>
            </Grid>
          </Grid>
        </Grid>
      </Grid>
    </Grid>
  );
})));
