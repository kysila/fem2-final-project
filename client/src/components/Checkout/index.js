import React, { useState } from 'react';
import { connect } from 'react-redux';
import {
  Grid, Typography, Button, Divider, Box, useMediaQuery, withTheme,
} from '@material-ui/core';
import {
  Route, Switch, Redirect, withRouter,
} from 'react-router-dom';
import { Logo } from '../../commons/Logo/Logo';
import { Steps } from '../../commons/Steps/Steps';
import { InputField } from '../../commons/InputField/InputField';
import { Checkout } from './Checkout';
import { Shipping } from './Shipping';
import { Payment } from './Payment';
import { CheckoutOrderProduct } from './CheckoutOrderProduct';
import { ShowOrderSummary } from './ShowOrderSummary';
import { useStyles } from './styles';


import MuseoSans from '../../fonts/MuseoSans-500.woff';

const museo = {
  fontFamily: 'Museo Sans 500',
  fontStyle: 'normal',
  src: `
    local('Museo Sans 500'),
    url(${MuseoSans}) format('woff')
  `,
};

const mapStateToProps = ({ cartReducer: { cart: { products } } }) => ({
  products: products.map(({
    cartQuantity: quantity,
    product: {
      name: title,
      imageUrls: [src],
      currentPrice: price,
      _id: key,
    },
  }) => ({
    key, src, title, price, quantity,
  })),
});

const CheckoutProcess = connect(mapStateToProps)(withRouter(withTheme((props) => {
  const classes = useStyles(props);
  const tablet = useMediaQuery(props.theme.breakpoints.down(1024));
  const mobile = useMediaQuery(props.theme.breakpoints.down(768));

  const [discount, setDiscount] = useState({
    value: '',
    variants: {
      KDT3X: 3,
      MFNM8: 5,
      B43T9: 10,
      BM2MB: 15,
      X4R9G: 20,
    },
    error: null,
  });

  const [discountApplied, setApplied] = useState(false);

  const steps = [
    {
      key: 'id-0',
      isActive: true,
      text: 'Cart',
    },
    {
      key: 'id-1',
      isActive: true,
      text: 'Information',
    },
    {
      key: 'id-2',
      isActive: props.location.pathname === '/checkout/shipping' || props.location.pathname === '/checkout/payment',
      text: 'Shipping',
    },
    {
      key: 'id-3',
      isActive: props.location.pathname === '/checkout/payment',
      text: 'Payment',
    },
  ];

  const calculatePrice = () => {
    const fullPrice = props.products
      .reduce((total, { price, quantity }) => Number(total) + (Number(price) * quantity), 0);
    const percent = discount.variants[discount.value] || 0;

    return {
      discountApplied,
      fullPrice: fullPrice.toFixed(2),
      percent,
      discountedPrice: discountApplied
        ? (fullPrice - (fullPrice * (percent / 100))).toFixed(2)
        : null,
    };
  };

  const calculated = calculatePrice();

  const changeDiscount = ({ target: { value } }) => {
    setDiscount({
      ...discount,
      value,
    });
    setApplied(false);
  };

  if (!props.products || !props.products.length) {
    return (
      <Redirect to="/" />
    );
  }

  return (
    <Grid
      container
      direction="row"
      alignItems="center"
      justify="center"
      className={classes.wrapper}
    >
      <Grid
        className={classes.container}
        container
        direction="column"
        alignItems="flex-start"
        justify="flex-start"
      >
        <Logo />
        <Steps className={classes.steps} items={steps} />
        <Grid
          className={classes.body}
          container
          direction={tablet ? 'column' : 'row'}
          alignItems={tablet && !mobile ? 'center' : mobile ? null : 'flex-start'}
          justify="center"
        >
          {
            tablet ? (
              <ShowOrderSummary
                products={props.products}
                calculatedPrice={calculated}
                applyDiscountHandler={() => setApplied(true)}
                DiscountProps={{
                  onChange: changeDiscount,
                  value: discount.value,
                }}
              />
            ) : null
          }
          <Grid
            item
            lg={tablet ? 12 : 7}
            md={tablet ? 12 : 7}
            xs={tablet ? 12 : 7}
          >
            { /* https://reacttraining.com/react-router/web/api/Switch */ }
            { /* https://tylermcginnis.com/react-router-nested-routes/ */ }
            <Route>
              <Switch>
                <Route exact path="/checkout/info">
                  <Checkout />
                </Route>
                <Route exact path="/checkout/shipping">
                  <Shipping />
                </Route>
                <Route exact path="/checkout/payment">
                  <Payment />
                </Route>
                <Redirect to="/checkout/info" />
              </Switch>
            </Route>
          </Grid>
          <Grid
            item
            lg={1}
            md={1}
            xs={1}
          />
          {
            !tablet ? (
              <Grid
                item
                lg={4}
                md={4}
                xs={4}
              >
                <Grid
                  container
                  direction="column"
                  alignItems="flex-start"
                  justify="flex-start"
                >
                  <Typography
                    style={{
                      fontSize: '16px',
                      lineHeight: '30px',
                      color: '#888888',
                      ...museo,
                    }}
                  >
                    Order Summary
                  </Typography>
                </Grid>
                {props.products.map((item) => <CheckoutOrderProduct {...item} />)}
                <Divider className={classes.divider} />
                <InputField
                  id="discount"
                  type="text"
                  label="Discount code"
                  labelWidth={110}
                  value={discount.value}
                  onChange={changeDiscount}
                  FormControlProps={{
                    className: classes.dividerFormControl,
                  }}
                  LabelProps={{
                    classes: {
                      root: classes.discountLabel,
                    },
                  }}
                  InputProps={{
                    classes: {
                      input: classes.discountInput,
                      adornedEnd: classes.discountAdornedEnd,
                    },
                    endAdornment: (
                      <Button
                        style={museo}
                        className={classes.discountButton}
                        variant="text"
                        onClick={() => {
                          setApplied(true);
                          localStorage.setItem('/checkout/discount', discount.value);
                        }}
                      >
                        Apply
                      </Button>
                    ),
                  }}
                />
                <Divider className={classes.divider} />
                <Grid
                  container
                  direction="column"
                  alignItems="center"
                  justify="center"
                >
                  <Grid
                    container
                    direction="row"
                    alignItems="center"
                    justify="space-between"
                  >
                    <Typography className={classes.afterDiscount} style={museo}>
                      Subtotal
                    </Typography>
                    <Typography className={classes.afterDiscount} style={museo}>
                      $
                      {
                        calculated.fullPrice
                      }
                    </Typography>
                  </Grid>
                  <Grid
                    container
                    direction="row"
                    alignItems="center"
                    justify="space-between"
                  >
                    <Typography className={classes.afterDiscount} style={museo}>
                      Shipping
                    </Typography>
                    <Typography className={classes.afterDiscount} style={museo}>
                      $0.00
                    </Typography>
                  </Grid>
                  {
                    calculated.percent > 0 && discountApplied ? (
                      <Grid
                        container
                        direction="row"
                        alignItems="center"
                        justify="space-between"
                      >
                        <Typography className={classes.afterDiscount} style={museo}>
                          Discount
                        </Typography>
                        <Typography className={classes.afterDiscount} style={museo}>
                          {
                            calculated.percent
                          }
                          %
                        </Typography>
                      </Grid>
                    ) : null
                  }
                </Grid>
                <Divider className={classes.divider} />
                <Grid
                  container
                  direction="row"
                  alignItems="center"
                  justify="space-between"
                >
                  <Typography className={classes.total} style={museo}>
                    Total
                  </Typography>
                  <Typography className={classes.total} style={museo}>
                    {
                      discountApplied && calculated.percent > 0 ? (
                        <React.Fragment>
                          <Typography component="span" className={classes.total} style={{ textDecoration: 'line-through', color: '#888888', ...museo }}>
                            $
                            {
                              calculated.fullPrice
                            }
                          </Typography>
                          {' / '} {/* eslint-disable-line */}
                        </React.Fragment>
                      ) : null
                    }
                    $
                    {
                      discountApplied && calculated.percent > 0
                        ? calculated.discountedPrice
                        : calculated.fullPrice
                    }
                  </Typography>
                </Grid>
              </Grid>
            ) : null
          }
        </Grid>
      </Grid>
      <Box
        className={classes.footer}
      >
        <Typography style={{ color: '#888888', fontSize: '12px', ...museo }}>
          Copyright © 2019 Electra.
        </Typography>
      </Box>
    </Grid>
  );
})));

export {
  CheckoutProcess as Checkout,
  CheckoutOrderProduct,
};
