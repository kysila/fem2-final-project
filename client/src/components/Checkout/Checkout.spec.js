/* eslint-disable */
import React from 'react';
import {shallow} from 'enzyme';
import {Checkout} from './Checkout';

describe('Checkout component', () => {
  it('Checkout render', () => {
    const rendered = shallow(
      <Checkout />,
    );

    expect(rendered).toMatchSnapshot();
  });
});
