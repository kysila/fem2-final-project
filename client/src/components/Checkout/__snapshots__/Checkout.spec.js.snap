// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout component Checkout render 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
